### PR TITLE
ns.getBitNodeMultipliers will now optionally accept 2 args

### PR DESCRIPTION
**Pull request by DaSpaceCat**
# ns.getBitNodeMultipliers will now optionally accept 2 args

MISC: `ns.getBitNodeMultipliers()` can now optionally accept 2 args, so that the player can get the multipliers of a bitnode that they are not currently in.

made per this request: https://discord.com/channels/415207508303544321/415207923506216971/1028774572436230244

### CODE REVIEW

File: src/ScriptEditor/NetscriptDefinitions.d.ts [hunk 1]
```diff
@@ -6603,9 +6603,11 @@ export interface NS {
    * print(ServerMaxMoney);
    * print(HackExpGain);
    * ```
-   * @returns Object containing the current BitNode multipliers.
+   * @param BitNode - BitNode you want the multipliers for.
+   * @param lvl - The level of the BitNode (for BN12)
+   * @returns Object containing the current (or supplied) BitNode multipliers.
    */
-  getBitNodeMultipliers(): BitNodeMultipliers;
+  getBitNodeMultipliers(BitNode, lvl): BitNodeMultipliers;
```
Comment (Snarling) [inline src/ScriptEditor/NetscriptDefinitions.d.ts:6610]: This documentation should show the types for these params and should show that they are optional. Variable names should not generally be capitalized.

```ts
  /* ...
   * @param bitNode - The BitNode to get multipliers for. Defaults to the current BitNode.
   * @param lvl - The level of the BitNode (for BN12). Defaults to 1.
   * @returns Object containing the current (or supplied) BitNode multipliers. */
getBitNodeMultipliers(bitNode?: number, lvl?: number): BitNodeMultipliers;
```

File: src/NetscriptFunctions.ts [hunk 2]
```diff
@@ -1072,13 +1073,32 @@ const base: InternalAPI<NS> = {
...
-  getBitNodeMultipliers: (ctx: NetscriptContext) => (): IBNMults => {
-    if (Player.sourceFileLvl(5) <= 0 && Player.bitNodeN !== 5) {
-      throw helpers.makeRuntimeErrorMsg(ctx, "Requires Source-File 5 to run.");
-    }
-    const copy = Object.assign({}, BitNodeMultipliers);
-    return copy;
-  },
+  getBitNodeMultipliers:
+    (ctx: NetscriptContext) =>
+    (_bitNode: unknown = Player.bitNodeN, _lvl: unknown = 1): IBNMults => {
```
Comment (Snarling) [inline src/NetscriptFunctions.ts:1078]: This default _lvl = 1 will result in an issue if the player is in a later iteration of BN12 and tries running getBitNodeMultipliers() with no arguments. Player would expect their current multipliers, they will instead only see the multipliers for 12.1

**Review by Snarling — CHANGES_REQUESTED** [review]

File: src/NetscriptFunctions.ts (hunk 2 above)
Comment (DaSpaceCat) [reply src/NetscriptFunctions.ts:1078]: resolved on lines 1097-1100

**Comment by quacksouls** [comment]
The formatter gives me this:
```sh
$ npm run format && git status

> bitburner@2.1.0 format
> prettier --write .

<Removed lots of verbose messages.>

On branch pr/4223
Your branch is up to date with 'github-desktop-DaSpaceCat/dev'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   src/NetscriptFunctions.ts

no changes added to commit (use "git add" and/or "git commit -a")

$ git diff src/NetscriptFunctions.ts
diff --git a/src/NetscriptFunctions.ts b/src/NetscriptFunctions.ts
index 7ba07497..d6b87326 100644
--- a/src/NetscriptFunctions.ts
+++ b/src/NetscriptFunctions.ts
@@ -1097,7 +1097,7 @@ const base: InternalAPI<NS> = {
         if (Player.bitNodeN == 12) {
           const copy = Object.assign({}, BitNodeMultipliers);
           return copy;
-        }
+        }
         const copy = Object.assign({}, specificBNMultipliers(bitNode, lvl));
         return copy;
       }
```

The documentation fails to build:
```sh
$ npm run doc

> bitburner@2.1.0 doc
> npx api-extractor run && npx api-documenter markdown && rm input/bitburner.api
.json && rm -r input


api-extractor 7.19.4  - https://api-extractor.com/

Using configuration from ./api-extractor.json
Analysis will use the bundled TypeScript version 4.5.5

ERROR: Internal Error: Unable to follow symbol for "const"
You have encountered a software defect. Please consider reporting the issue to t
he maintainers of this application.
```
This is because your branch does not include the fix for the doc builder at commit f803355335f77624ed64d21268db46fc76c50ceb.  That's OK.  A rebase should fix the problem.  I rebased the PR on top of the current `dev` branch at 

https://github.com/danielyxie/bitburner/tree/dev

and the doc builds fine.

**Comment by DaSpaceCat** [comment]
alright, committed the formatting issue fix. I tried rebasinf but it said the dev branch was up to date, do I need to do something with that?

File: src/NetscriptFunctions.ts [hunk 3]
```diff
@@ -1072,13 +1073,36 @@ const base: InternalAPI<NS> = {
...
+      }
+
+      if (lvl < 0 && lvl != undefined) {
+        throw helpers.makeRuntimeErrorMsg(ctx, "Invalid level, must be an integer above 0.");
+      }
+
+      if (bitNode != 12) {
+        const copy = Object.assign({}, specificBNMultipliers(bitNode, 0));
+        return copy;
+      } else if (bitNode == 12) {
```
Comment (Snarling) [inline src/NetscriptFunctions.ts:1096]: This section currently will always return the current multipliers if the player specifies BN12 from within BN12. e.g. if I'm on 12.4 and I'm wondering what the multipliers are going to be like at 12.8, I can't check that.

**Review by Snarling — CHANGES_REQUESTED** [review]
I think this is getting overcomplicated by the way we're currently handling default values. Defaults in params makes it impossible to determine if the player actually provided a value, so it makes it impossible for us to vary function behavior based on whether something was provided. I took the defaults out of parameters and moved some checks around in a way that simplifies things. I also added checks for the provided bitNode / level actually being integers since there was nothing enforcing those requirements. Feel free to modify the below however you need to.

```ts
  getBitNodeMultipliers: (ctx: NetscriptContext) => (_bn: unknown, _lvl: unknown): IBNMults => {
    if (Player.sourceFileLvl(5) <= 0 && Player.bitNodeN !== 5) throw helpers.makeRuntimeErrorMsg(ctx, "Requires Source-File 5 to run.");
    //Return the current multipliers if this was called with no args.
    if (_bn === undefined) return Object.assign({}, BitNodeMultipliers);

    const bitNode = helpers.number(ctx, "bitNode", _bn);
    // Current max BN is 13. Adjust as more BN are added.
    if (bitNode > 13 || bitNode < 1 || bitNode !== Math.floor(bitNode)) {
      throw helpers.makeRuntimeErrorMsg(ctx, `Provided bitNode (${bitNode}) invalid. Must be an integer between 1-13`);
    }

    // BN12 cares about level, so it has separate logic. Any new BN that cares about level should be added to this block.
    if (bitNode === 12) {
      if (_lvl === undefined) throw helpers.makeRuntimeErrorMsg(ctx, `Level was not specified. The requested BitNode: ${bitNode} requires a specified level.`)
      const level = helpers.number(ctx, "level", _lvl);
      if (level <= 0 || level !== Math.floor(level)) throw helpers.makeRuntimeErrorMsg(ctx, `Provided level (${level}) invalid. Must be an integer greater than 0.`);
      return Object.assign({}, specificBNMultipliers(bitNode, level));
    }

    // Just use level 1 for bitNodes that don't care about level.
    return Object.assign({}, specificBNMultipliers(bitNode, 1));
  },
```

File: src/NetscriptFunctions.ts (hunk 3 above)
Comment (Snarling) [reply src/NetscriptFunctions.ts:1096]: I highlighted the wrong lines here, it should be 1096-1100

**Comment by Snarling** [comment]
> alright, committed the formatting issue fix. I tried rebasing, but it said the dev branch was up to date, do I need to do something with that?

@DaSpaceCat @quacksouls Since this PR doesn't conflict with dev it doesn't actually matter if it gets rebased or not. If the doc is building fine after rebase, then it will build fine after merging as well.

But here's how you would rebase on GH desktop (with your branch selected, try to rebase into the upstream dev branch). You may need to switch to the dev branch and "fetch upstream" first if it's telling you you're up to date.

image

**Comment by quacksouls** [comment]
> alright, committed the formatting issue fix.

Yay `\o/`

> I tried rebasing, but it said the dev branch was up to date, do I need to do something with that?

You don't need to do anything.  On this PR conversation, it says

`This branch has no conflicts with the base branch`

so no rebase necessary.

**Comment by DaSpaceCat** [comment]
just fixed the BN12 issues with the code @Snarling suggested, ran the format, and tested it.

**Review by Snarling — APPROVED** [review]

**Comment by hydroflame** [comment]
Merged separatly

**Pull request closed without merging**